Globbing tests

ruchim opened the pull request:
- Test string interpolation inside of the glob() function
- Test globbing boundaries from different directories

ruchim committed:
globbing behavior tests
test globing outside cromwell dir

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,66 @@
+}
+
+task countArrays {
+Array[String] in1
```
indent

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,66 @@
+task writeFiles {
+String dir
```
indent

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,66 @@
+  >>>
+  output {
+    Array[File] strInterpolation_1 = glob(dir + "/*.txt")
+    Array[File] strInterpolation_2=  glob("${dir}/*.txt")
```
Only `_2` is interpolation. `_1` is just the String `+` operator.

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,66 @@
+    Array[File] strInterpolation_2=  glob("${dir}/*.txt")
+
+    Array[File] callDirectory = glob("*.txt")
+    Array[File] specificDirectory = glob("out/*.txt")
```
Should this be `glob(dir + "/*.txt")`?

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,66 @@
+ String outputDir = "out"
+
+ call writeFiles { input: dir = outputDir }
+ call countArrays {input:in1 = writeFiles.strInterpolation_1,
```
ToL: It's a little embarrassing that this isn't an engine function 😳

cjllanwarne replied:
ToL2: It would be nice if this workflow listed the file contents so we could have a more accurate output validation. Maybe we could do this instead of the `count_arrays` call?
```
scatter (f in writeFiles.strInterpolation_1) {
  String strInterpolation_1_contents = read_string(f)
}
# And for the other arrays ...
# ...
output {
  strInterpolation_1_contents
  # And the others
}
```

ruchim replied:
^ like this suggestion a lot. Originally I was trying to match file paths and they were absolute and it was a mess with the glob-dir, but this is better

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,66 @@
+ output {
+  countArrays.*
+ }
+}
```
Newline

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior.test:
```diff
@@ -0,0 +1,12 @@
+  workflowName: globbingBehavior
+  status: Succeeded
+  "outputs.globbingBehavior_countArrays_out": "[2, 2, 3, 2]"
+}
```
Newline

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior.test:
```diff
@@ -0,0 +1,12 @@
+metadata {
+  workflowName: globbingBehavior
+  status: Succeeded
+  "outputs.globbingBehavior_countArrays_out": "[2, 2, 3, 2]"
```
ToL (as below) if this is possible to make into a real validation, that'd be awesome...

kshakir committed:
Added a test to make sure "|" doesn't get stripped from commands.

kshakir committed:
Updated instructions on how to run a camel case test tag.

kshakir committed:
Fixed expected outputs for globbingBehavior.
Slight mod to test to check for cd'ing to a sub directories.
Added test for globbing of chains of symlinks.

cjllanwarne commented on src/main/resources/standardTestCases/dont_strip_line_prefix/dont_strip_line_prefix.wdl:
```diff
@@ -0,0 +1,19 @@
+workflow wf_prefix {
+  call prefix
+  output {
+     prefix.out
```
Since this is a new file, I'd prefer a new style output block (e.g. `String prefix_out = prefix.out`)

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,74 @@
+   strInterpolation_contents
+   callDirectory_contents
+   specificDirectory_contents
+   easyGlob_contents
```
Same request for new style outputs here

cjllanwarne commented on src/main/resources/standardTestCases/globbingBehavior/globbingBehavior.wdl:
```diff
@@ -0,0 +1,74 @@
+  call unexpectedReturnCode {}
+
+  scatter (f in writeFiles.strInterpolation) {
+    String strInterpolation_contents = read_string(f)
```
Do we really want to officially endorse "you can use 'scatter' as a 'map' function" in our test cases?

cjllanwarne reviewed (changes requested):
ToL: I think it's important to check that symlinks are correctly collected across backends too.

E.g. if you list a symlink as a File output you'll get the dereferenced file, not the link. The same should be true if you have a directory of symlinks.

(And even chains of symlinks)

mcovarr commented:
:+1:

[![Approved with PullApprove](https://img.shields.io/badge/pullapprove-approved-brightgreen.svg)](https://pullapprove.com/broadinstitute/centaur/pull-request/124/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/centaur)

cjllanwarne commented:
👍 

[![Approved with PullApprove](https://img.shields.io/badge/pullapprove-approved-brightgreen.svg)](https://pullapprove.com/broadinstitute/centaur/pull-request/124/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/centaur)
